lib/puppet/util/windows/principal.rb

```diff
@@ -0,0 +1,177 @@
+require 'puppet/util/windows'
+
+module Puppet::Util::Windows::SID
+  class Principal
+    extend FFI::Library
+    attr_reader :account, :sid_bytes, :sid, :domain, :domain_account, :account_type
+
+    def initialize(account, sid_bytes, sid, domain, account_type)
+      @account = account
+      @sid_bytes = sid_bytes
+      @sid = sid
+      @domain = domain
+      @domain_account = domain && !domain.empty? ?
+        "#{domain}\\#{account}" : account
+
+      @account_type = account_type
+    end
+
+    # added for backward compatibility
+    def ==(compare)
+      compare.is_a?(Puppet::Util::Windows::SID::Principal) &&
+        @sid_bytes == compare.sid_bytes
+    end
+
+    # added for backward compatibility
+    def to_s
+      @sid
+    end
+
+    # = 8 + max sub identifiers (15) * 4
+    MAXIMUM_SID_BYTE_LENGTH = 68
+
+    ERROR_INSUFFICIENT_BUFFER = 122
+
+    def self.lookup_account_name(system_name = nil, account_name)
+      system_name_ptr = FFI::Pointer::NULL
+      begin
+        if system_name
+          system_name_wide = Puppet::Util::Windows::String.wide_string(system_name)
+          # uchar here is synonymous with byte
+          system_name_ptr = FFI::MemoryPointer.new(:byte, system_name_wide.bytesize)
+          system_name_ptr.put_array_of_uchar(0, system_name_wide.bytes.to_a)
+        end
+
+        FFI::MemoryPointer.from_string_to_wide_string(account_name) do |account_name_ptr|
+          FFI::MemoryPointer.new(:byte, MAXIMUM_SID_BYTE_LENGTH) do |sid_ptr|
+            FFI::MemoryPointer.new(:dword, 1) do |sid_length_ptr|
+              FFI::MemoryPointer.new(:dword, 1) do |domain_length_ptr|
+                FFI::MemoryPointer.new(:uint32, 1) do |name_use_enum_ptr|
+
+                sid_length_ptr.write_dword(MAXIMUM_SID_BYTE_LENGTH)
+                success = LookupAccountNameW(system_name_ptr, account_name_ptr, sid_ptr, sid_length_ptr,
+                  FFI::Pointer::NULL, domain_length_ptr, name_use_enum_ptr)
+                last_error = FFI.errno
+
+                if (success == FFI::WIN32_FALSE && last_error != ERROR_INSUFFICIENT_BUFFER)
+                  raise Puppet::Util::Windows::Error.new('Failed to call LookupAccountNameW', last_error)
```

Comment:
If a valid SID that does not exist last_error is 1332 (No mapping between account....) whereas is you use a byte array that is not a valid SID e.g. [0] or [5] (anything but [1]), last_error is 87 which is invalid parameter.  Should the two scenarios be surfaced differently (Error code 1332 meant the API was called successfully but the SID supplied didn't exist)